File: Assets/Scripts/APlus/ExamSession.cs
using System;
using System.Collections.Generic;

namespace APlus
{
    public enum SessionPhase { NotStarted, Running, Died, Submitted }

    /// <summary>一场考试的可调参数。三条推进线（异象强度 / 报时污染 / 死亡退化）刻意错峰（2.4 丝滑技术 4）。</summary>
    [Serializable]
    public sealed class SessionConfig
    {
        public int RoundSize = 9;
        /// <summary>4.3：推荐 3,3,3 + 抖动 1（Docs/难度配比标定.md §5）。</summary>
        public int MixEasy = 3, MixMedium = 3, MixHard = 3;
        public AllocationMode Mode = AllocationMode.QuotaJitter;
        public int Jitter = 1;

        /// <summary>6.2：异象强度随累计死亡在这个区间里从 0 爬到 1。</summary>
        public DeathWindow AnomalyIntensity = new DeathWindow(3, 15);
        /// <summary>3.4：第几次模拟考试起，报时开始变慢 / 倒退 / 改念考号。</summary>
        public int SlowFromAttempt = 4;
        public int RewindFromAttempt = 8;
        public int RosterFromAttempt = 12;

        /// <summary>A2：听力期间离开卷子多久算违规。</summary>
        public float ListeningViolationGrace = 0.6f;

        public MixSpec Mix { get { return new MixSpec("session", MixEasy, MixMedium, MixHard); } }

        public BroadcastTamper TamperFor(int attempt)
        {
            if (attempt >= RosterFromAttempt) return BroadcastTamper.Roster;
            if (attempt >= RewindFromAttempt) return BroadcastTamper.Rewind;
            if (attempt >= SlowFromAttempt) return BroadcastTamper.Slow;
            return BroadcastTamper.Normal;
        }
    }

    /// <summary>一帧里发生的事。呈现层据此放声音、翻记名册、起死亡表现。</summary>
    public sealed class SessionEvents
    {
        public readonly List<BroadcastCue> Broadcasts = new List<BroadcastCue>();
        public readonly List<TickResult> Gaze = new List<TickResult>();
        public bool Died;
        public bool TimeUp;

        public void Clear()
        {
            Broadcasts.Clear();
            Gaze.Clear();
            Died = false;
            TimeUp = false;
        }
    }

    /// <summary>
    /// 一场考试的全部逻辑：抽卷 → 注视调度 → 计时与广播 → 作答 → 交卷或死亡 → 写回 RunState。
    /// 不依赖 UnityEngine；Unity 侧的 MonoBehaviour 每帧调 Tick，并把输入转成 SetGaze / Select。
    /// </summary>
    public sealed class ExamSession
    {
        public readonly GameData Data;
        public readonly StringTable Table;
        public readonly RunState Run;
        public readonly InvigilatorProfile Profile;
        public readonly SessionConfig Config;
        public readonly DeathCurve Curve;

        public SessionPhase Phase { get; private set; }
        public RoundPlan Plan { get; private set; }
        public AnswerContext Context { get; private set; }
        public List<ExamQuestion> Questions { get; private set; }
        public GazeStateMachine Gaze { get; private set; }
        public GazeDirector Director { get; private set; }
        public ExamClock Clock { get; private set; }
        public BroadcastTamper Tamper { get; private set; }
        public DeathPresentation Death { get; private set; }

        float _listeningOffPaper;
        bool _listeningRecorded;

        public ExamSession(GameData data, StringTable table, RunState run, InvigilatorProfile profile, SessionConfig config, DeathCurve curve)
        {
            Data = data;
            Table = table;
            Run = run;
            Profile = profile ?? InvigilatorProfile.Podium();
            Config = config ?? new SessionConfig();
            Curve = curve ?? new DeathCurve();
            Questions = new List<ExamQuestion>();
            Phase = SessionPhase.NotStarted;
        }

        public void Begin()
        {
            ExamRoundGenerator gen = new ExamRoundGenerator(Data);
            gen.Mode = Config.Mode;
            gen.Jitter = Config.Jitter;
            Plan = gen.BuildRound(Run.RoundSeed, Config.RoundSize, Config.Mix);

            Context = AnswerContext.FromRun(Run, Plan.Items.Count, Table);
            Questions = new List<ExamQuestion>();
            for (int i = 0; i < Plan.Items.Count; i++)
            {
                ExamQuestion q = ChoiceBuilder.Build(Plan.Items[i].Chain, Context);
                q.Plan = Plan.Items[i];
                Questions.Add(q);
            }

            Gaze = new GazeStateMachine();
            Director = new GazeDirector(Profile, unchecked(Run.RoundSeed ^ 0x2545F491), Config.AnomalyIntensity.Ramp(Run.TotalDeaths));

            Clock = new ExamClock();
            Tamper = Config.TamperFor(Run.Attempt);
            string[] roster = new string[Clock.RemainingAnnouncements.Length];
            for (int i = 0; i < roster.Length; i++) roster[i] = Run.RosterNumber(i);
            Clock.RosterNumbers = roster;
            Clock.Build(Tamper);

            _listeningOffPaper = 0f;
            _listeningRecorded = false;
            Phase = SessionPhase.Running;
        }

        public void SetGaze(GazeState state)
        {
            if (Phase == SessionPhase.Running) Gaze.SetState(state);
        }

        /// <summary>作答只能在「卷子」状态（2.1：这是安全区，也是唯一能写字的地方）。</summary>
        public bool Select(int questionIndex, int choiceIndex)
        {
            if (Phase != SessionPhase.Running || Gaze.State != GazeState.Paper) return false;
            if (questionIndex < 0 || questionIndex >= Questions.Count) return false;
            ExamQuestion q = Questions[questionIndex];
            if (choiceIndex < 0 || choiceIndex >= q.Choices.Length) return false;
            q.Selected = choiceIndex;
            return true;
        }

        public void Tick(float dt, SessionEvents events)
        {
            if (Phase != SessionPhase.Running) return;
            if (dt < 0f) dt = 0f;

            Clock.Tick(dt, events.Broadcasts);
            Director.Listening = Clock.InListening;
            Director.Tick(dt, Gaze);

            TickResult r = Gaze.Tick(dt);
            if (r != TickResult.Idle && r != TickResult.Safe) events.Gaze.Add(r);

            if (!Gaze.IsDead) TickListeningRule(dt, events);

            if (Gaze.IsDead)
            {
                Phase = SessionPhase.Died;
                Death = Curve.Evaluate(Run.NextDeathOrdinal);
                events.Died = true;
                return;
            }
            if (Clock.IsOver)
            {
                events.TimeUp = true;
                Submit();
            }
        }

        void TickListeningRule(float dt, SessionEvents events)
        {
            if (!Clock.InListening || _listeningRecorded) return;
            if (Gaze.State == GazeState.Paper) { _listeningOffPaper = 0f; return; }
            _listeningOffPaper += dt;
            if (_listeningOffPaper < Config.ListeningViolationGrace) return;
            _listeningRecorded = true;
            events.Gaze.Add(Gaze.RecordViolation());
        }

        public void Submit()
        {
            if (Phase == SessionPhase.Running) Phase = SessionPhase.Submitted;
        }

        public int Score
        {
            get
            {
                int n = 0;
                for (int i = 0; i < Questions.Count; i++) if (Questions[i].IsCorrect) n++;
                return n;
            }
        }

        /// <summary>把这一场的结果写回 RunState，返回下一步去哪。只能调一次。</summary>
        public RunTransition Finish()
        {
            if (Phase != SessionPhase.Died && Phase != SessionPhase.Submitted)
                throw new InvalidOperationException("session is still " + Phase);
            bool died = Phase == SessionPhase.Died;
            Phase = SessionPhase.NotStarted;
            return Run.CompleteAttempt(died, Gaze.TimesRecorded, Gaze.Cause);
        }
    }
}

File: src/core/examSession.ts
import { AnswerContext } from './answerExpression';
import { buildChoice, type ExamQuestion } from './choiceBuilder';
import { ClassroomLayout } from './classroomLayout';
import { DeathCurve, DeathWindow, type DeathPresentation } from './deathCurve';
import { BroadcastTamper, ExamClock, type BroadcastCue } from './examClock';
import type { GameData } from './gameData';
import { GazeDirector, InvigilatorProfile } from './gazeDirector';
import { FlipPhone, PhoneEventKind, type PhoneEvent, type PhoneKey } from './flipPhone';
import { GazeState, GazeStateMachine, TickResult } from './gazeStateMachine';
import { InvigilatorPatrol, PatrolEventKind, PatrolProfile, type PatrolEvent } from './invigilatorPatrol';
import { AllocationMode, ExamRoundGenerator, MixSpec, type RoundPlan } from './roundGenerator';
import type { RunState, RunTransition } from './runState';
import type { StringTable } from './stringTable';

export const SessionPhase = { NotStarted: 'not_started', Running: 'running', Died: 'died', Submitted: 'submitted' } as const;
export type SessionPhase = (typeof SessionPhase)[keyof typeof SessionPhase];

/** 一场考试的可调参数。三条推进线（异象强度 / 报时污染 / 死亡退化）刻意错峰（2.4 丝滑技术 4）。 */
export class SessionConfig {
  roundSize = 9;
  /** 4.3：推荐 3,3,3 + 抖动 1（Docs/难度配比标定.md §5）。 */
  mixEasy = 3;
  mixMedium = 3;
  mixHard = 3;
  mode: AllocationMode = AllocationMode.QuotaJitter;
  jitter = 1;

  /** 6.2：异象强度随累计死亡在这个区间里从 0 爬到 1。 */
  anomalyIntensity = new DeathWindow(3, 15);
  /** 3.4：第几次模拟考试起，报时开始变慢 / 倒退 / 改念考号。 */
  slowFromAttempt = 4;
  rewindFromAttempt = 8;
  rosterFromAttempt = 12;

  /** A2：听力期间离开卷子多久算违规。 */
  listeningViolationGrace = 0.6;

  get mix(): MixSpec {
    return new MixSpec('session', this.mixEasy, this.mixMedium, this.mixHard);
  }

  tamperFor(attempt: number): BroadcastTamper {
    if (attempt >= this.rosterFromAttempt) return BroadcastTamper.Roster;
    if (attempt >= this.rewindFromAttempt) return BroadcastTamper.Rewind;
    if (attempt >= this.slowFromAttempt) return BroadcastTamper.Slow;
    return BroadcastTamper.Normal;
  }
}

/** 一帧里发生的事。呈现层据此放声音、翻记名册、起死亡表现。 */
export class SessionEvents {
  readonly broadcasts: BroadcastCue[] = [];
  readonly gaze: TickResult[] = [];
  readonly phone: PhoneEvent[] = [];
  readonly patrol: PatrolEvent[] = [];
  died = false;
  timeUp = false;
  /** 巡考这一帧抓了你现行（她已经闪现到你身边）。 */
  caught = false;

  clear(): void {
    this.broadcasts.length = 0;
    this.gaze.length = 0;
    this.phone.length = 0;
    this.patrol.length = 0;
    this.died = false;
    this.timeUp = false;
    this.caught = false;
  }
}

/**
 * 一场考试的全部逻辑：抽卷 → 注视调度 → 计时与广播 → 作答 → 交卷或死亡 → 写回 RunState。
 * 不依赖渲染引擎；呈现层每帧调 tick，并把输入转成 setGaze / select / pressPhoneKey。
 *
 * 巡考模式（InvigilatorProfile.patrol，决策 #43）：师视不再按时间表落下，而是由在过道里走动的
 * InvigilatorPatrol 用视锥和耳朵判定；被她抓现行 = 记名一次，她闪现到你身边瞪着你。
 */
export class ExamSession {
  readonly profile: InvigilatorProfile;
  readonly config: SessionConfig;
  readonly curve: DeathCurve;
  readonly layout: ClassroomLayout;
  readonly patrolProfile: PatrolProfile;

  private _phase: SessionPhase = SessionPhase.NotStarted;
  plan!: RoundPlan;
  context!: AnswerContext;
  questions: ExamQuestion[] = [];
  gaze!: GazeStateMachine;
  director!: GazeDirector;
  clock!: ExamClock;
  tamper: BroadcastTamper = BroadcastTamper.Normal;
  death: DeathPresentation | null = null;
  /** 决策 #42：答案都在这台翻盖机里。 */
  phone!: FlipPhone;
  /** 巡考老师；podium 模式下为 null（师视由 GazeDirector 按时间表注入）。 */
  patrol: InvigilatorPatrol | null = null;

  private listeningOffPaper = 0;
  private listeningRecorded = false;
  private lookingAtTeacher = false;
  private readonly phoneQueue: PhoneEvent[] = [];

  constructor(
    readonly data: GameData,
    readonly table: StringTable,
    readonly run: RunState,
    profile?: InvigilatorProfile | null,
    config?: SessionConfig | null,
    curve?: DeathCurve | null,
    layout?: ClassroomLayout | null,
    patrolProfile?: PatrolProfile | null,
  ) {
    this.profile = profile ?? InvigilatorProfile.podium();
    this.config = config ?? new SessionConfig();
    this.curve = curve ?? new DeathCurve();
    this.layout = layout ?? new ClassroomLayout();
    this.patrolProfile = patrolProfile ?? new PatrolProfile();
  }

  get phase(): SessionPhase { return this._phase; }

  begin(): void {
    const gen = new ExamRoundGenerator(this.data);
    gen.mode = this.config.mode;
    gen.jitter = this.config.jitter;
    this.plan = gen.buildRound(this.run.roundSeed, this.config.roundSize, this.config.mix);

    this.context = AnswerContext.fromRun(this.run, this.plan.items.length, this.table);
    this.questions = this.plan.items.map((item) => {
      const q = buildChoice(item.chain, this.context);
      q.plan = item;
      return q;
    });

    this.gaze = new GazeStateMachine();
    this.director = new GazeDirector(this.profile, this.run.roundSeed ^ 0x2545f491, this.config.anomalyIntensity.ramp(this.run.totalDeaths));

    this.clock = new ExamClock();
    this.tamper = this.config.tamperFor(this.run.attempt);
    this.clock.rosterNumbers = this.clock.remainingAnnouncements.map((_, i) => this.run.rosterNumber(i));
    this.clock.build(this.tamper);

    this.phone = FlipPhone.forSession(this.questions, (key) => this.table.get(key), this.run.roundSeed ^ 0x7f4a7c15, this.run.attempt);
    this.phoneQueue.length = 0;
    this.patrol = this.profile.teacherScheduled ? null : new InvigilatorPatrol(this.layout, this.patrolProfile, this.run.roundSeed ^ 0x3c6ef372);
    this.gaze.summonOnExposure = this.patrol === null;
    this.lookingAtTeacher = false;

    this.death = null;
    this.listeningOffPaper = 0;
    this.listeningRecorded = false;
    this._phase = SessionPhase.Running;
  }

  /** 被抓现行的那一下（她闪现到你身边）：输入无效，视线被按回卷子上。 */
  get inputLocked(): boolean {
    return this.patrol?.busy ?? false;
  }

  setGaze(state: GazeState): void {
    if (this._phase !== SessionPhase.Running || this.inputLocked) return;
    const wasPhone = this.gaze.state === GazeState.Phone;
    this.gaze.setState(state);
    if (wasPhone && this.gaze.state !== GazeState.Phone) this.phone.putAway();
  }

  /** 呈现层每帧告诉我们：玩家是不是正对着老师的脸（对视判定）。 */
  setLookingAtTeacher(looking: boolean): void {
    this.lookingAtTeacher = looking;
  }

  /** 掏着手机时按一个键。返回 false = 手机没掏出来 / 没电 / 考试不在进行。 */
  pressPhoneKey(key: PhoneKey): boolean {
    if (this._phase !== SessionPhase.Running || this.inputLocked) return false;
    if (this.gaze.state !== GazeState.Phone || this.phone.dead) return false;
    this.phone.press(key, this.phoneQueue);
    return true;
  }

  /** 作答只能在「卷子」状态（2.1：这是安全区，也是唯一能写字的地方）。 */
  select(questionIndex: number, choiceIndex: number): boolean {
    if (this._phase !== SessionPhase.Running || this.gaze.state !== GazeState.Paper) return false;
    const q = this.questions[questionIndex];
    if (!q || choiceIndex < 0 || choiceIndex >= q.choices.length) return false;
    q.selected = choiceIndex;
    return true;
  }

  tick(dt: number, events: SessionEvents): void {
    if (this._phase !== SessionPhase.Running) return;
    if (dt < 0) dt = 0;

    this.clock.tick(dt, events.broadcasts);
    this.director.listening = this.clock.inListening;
    if (!this.inputLocked) this.director.tick(dt, this.gaze);

    const phoneStart = events.phone.length;
    events.phone.push(...this.phoneQueue);
    this.phoneQueue.length = 0;
    this.phone.tick(dt, this.gaze.state === GazeState.Phone, events.phone);
    if (this.patrol) this.tickPatrol(dt, events, phoneStart);

    if (!this.gaze.isDead) {
      const r = this.gaze.tick(dt);
      if (r !== TickResult.Idle && r !== TickResult.Safe) events.gaze.push(r);
    }

    if (!this.gaze.isDead) this.tickListeningRule(dt, events);

    if (this.gaze.isDead) {
      this._phase = SessionPhase.Died;
      this.death = this.curve.evaluate(this.run.nextDeathOrdinal);
      events.died = true;
      return;
    }
    if (this.clock.isOver) {
      events.timeUp = true;
      this.submit();
    }
  }

  private tickPatrol(dt: number, events: SessionEvents, phoneStart: number): void {
    const patrol = this.patrol!;
    let ringing = false;
    let keyTones = 0;
    for (let i = phoneStart; i < events.phone.length; i++) {
      const e = events.phone[i];
      if (e.kind === PhoneEventKind.Ring) ringing = true;
      else if (e.kind === PhoneEventKind.Key && e.audible) keyTones++;
    }
    const patrolStart = events.patrol.length;
    patrol.tick(dt, {
      gaze: this.gaze.state,
      lookingAtTeacher: this.lookingAtTeacher,
      exposure: this.gaze.exposure,
      ringing,
      keyTones,
      listening: this.clock.inListening,
    }, this.gaze, events.patrol, events.gaze);
    for (let i = patrolStart; i < events.patrol.length; i++) {
      if (events.patrol[i].kind !== PatrolEventKind.Caught) continue;
      events.caught = true;
      if (this.gaze.state === GazeState.Phone) this.phone.putAway();
      this.gaze.setState(GazeState.Paper);
    }
  }

  private tickListeningRule(dt: number, events: SessionEvents): void {
    if (!this.clock.inListening || this.listeningRecorded) return;
    if (this.gaze.state === GazeState.Paper) { this.listeningOffPaper = 0; return; }
    this.listeningOffPaper += dt;
    if (this.listeningOffPaper < this.config.listeningViolationGrace) return;
    this.listeningRecorded = true;
    events.gaze.push(this.gaze.recordViolation());
  }

  submit(): void {
    if (this._phase === SessionPhase.Running) this._phase = SessionPhase.Submitted;
  }

  get score(): number {
    return this.questions.filter((q) => q.isCorrect).length;
  }

  /** 把这一场的结果写回 RunState，返回下一步去哪。只能调一次。 */
  finish(): RunTransition {
    if (this._phase !== SessionPhase.Died && this._phase !== SessionPhase.Submitted) {
      throw new Error('session is still ' + this._phase);
    }
    const died = this._phase === SessionPhase.Died;
    this._phase = SessionPhase.NotStarted;
    return this.run.completeAttempt(died, this.gaze.timesRecorded, this.gaze.cause);
  }
}
